pom.xml

```diff
@@ -84,7 +85,7 @@
     <dependency>
       <groupId>org.jenkins-ci.plugins</groupId>
       <artifactId>structs</artifactId>
-      <version>1.17</version>
+      <version>1.18-rc265.27f956fa74ee</version> <!-- TODO https://github.com/jenkinsci/structs-plugin/pull/43 -->
```

Comment:
@MarkEWaite take note that this has not yet been reviewed, much less merged or released.

Reply:
I assume that once the structs plugin changes have been reviewed, built, tested, and released, you'll update this pull request to the new structs version.

I assume also that this fix will be needed on the stable-3.9 branch so that it does not need to wait for the release of git plugin 4.0.

Reply:
> I assume that once the structs plugin changes have been reviewed, built, tested, and released

If and when.

> you'll update this pull request to the new structs version

Yes.

> I assume also that this fix will be needed on the stable-3.9 branch

I have no opinions about backporting issues—that is up to the maintainer. I have no reason to believe #698 + this would _not_ `cherry-pick -x` cleanly.

Reply:
Though the `structs` API assumes Jenkins 2.60.x and Java 8. And I think Git 3.x still runs against obsolete LTS lines? So that would be a problem. The `structs` patch would itself be easy to cherry-pick, but the Jenkins plugin manager does not grok a dependency like `1.14.1+ | 1.18+`, so backporting an API would introduce the risk of someone updating `git` but leaving `structs 1.17` installed and getting a linkage error.

Reply:
Since it sounds like no general 4.x release (i.e., from `master`) is forthcoming, I would suggest either:

* Make the 3.9.x branch use at least Jenkins 2.60.x. There is no legitimate reason for anyone running an even more ancient LTS line (2.46.x or older) to be expecting any kind of plugin updates at this point. Even if you have CloudBees support, you are not still receiving security updates, so you are running with multiple unpatched vulnerabilities; for OSS users, I do not think the update center even advertises plugins for that line any more (IIRC it goes about five lines back).
* Failing that, create a 3.10.x branch forked from the current tip of 3.9.x but using a newer baseline, and `cherry-pick` some fixes from `master` to that, as well as continuing to `merge` up changes from 3.9.x.

Reply:
I prefer the second option (create `stable-3.10` branch forked from stable-3.9) to cover the major shift in Jenkins version requirement. Instead of requiring Jenkins 1.642, it will likely require Jenkins 2.121.1 or newer.

Reply:
In case you were wondering, `VersionNumber` does compare `3.10.1` > `3.9.4` etc. as you would want. (I.e., it is not just parsing as `double` or something.)

Reply:
Thanks @jglick!

I created the [stable-3.10 branch](https://github.com/jenkinsci/git-plugin/commits/stable-3.10) and have pushed changes to it to:
* require Jenkins 2.121.1 
* use the most recent parent pom
* deliver incremental artifacts
* test with Java 8 with Jenkins 2.121, Java 8 with Jenkins 2.164.1, and Java 11 with Jenkins 2.164.1 on Windows and Linux

Tests are passing in all those environments on my CI infrastructure.  I've also checked that the tests pass on multiple platforms (CentOS 6, CentOS 7, Debian 8, Debian 9, Debian Testing, FreeBSD 12, Ubuntu 16, Ubuntu 18, Windows).

I created a [disposable branch](https://github.com/MarkEWaite/git-plugin/commits/stable-3.10-JENKINS-51638) to include the changes in this pull request. When I try to build that branch, it fails the build with:

```
[INFO] --- access-modifier-checker:1.16:enforce (default-enforce) @ git ---
[ERROR] hudson/plugins/git/UserMergeOptions$DescriptorImpl:171 org/jenkinsci/plugins/structs/describable/CustomDescribableModel is still in beta
```

Any hints what mistake I'm making that causes compilation to fail in that branch?

Reply:
You did not backport

```xml
<useBeta>true</useBeta>
```